File: modules/collab/test/integration/membership.test.mjs
// Server-side membership source: strict JSON decoding + role resolution over verified users.

import { mkdtempSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import { join } from "node:path";
import { describe, expect, it } from "vitest";
import {
  createMembershipRoleResolver,
  decodeMemberships,
  loadMembershipRoleResolver,
} from "../../server/membership.mjs";

describe("membership source — decoding", () => {
  it("decodes a strict room -> subject -> role membership object", () => {
    const memberships = decodeMemberships({
      rooms: {
        "org_acme/claims": {
          "auth0|adjuster": "editor",
          "auth0|auditor": "viewer",
        },
      },
    });

    expect(memberships.get("org_acme/claims")?.get("auth0|adjuster")).toBe("editor");
    expect(memberships.get("org_acme/claims")?.get("auth0|auditor")).toBe("viewer");
  });

  it("rejects malformed membership data loudly", () => {
    expect(() => decodeMemberships({ rooms: { claims: { "auth0|u": "admin" } } })).toThrow(
      /invalid role/,
    );
    expect(() => decodeMemberships({ rooms: { claims: ["auth0|u"] } })).toThrow(
      /must be an object/,
    );
    expect(() => decodeMemberships({})).toThrow(/rooms object/);
  });
});

describe("membership source — role resolution", () => {
  it("grants the role listed for the verified subject", () => {
    const authorizeRoom = createMembershipRoleResolver({
      memberships: decodeMemberships({
        rooms: {
          "org_acme/claims": { "auth0|adjuster": "editor", "auth0|auditor": "viewer" },
        },
      }),
    });

    expect(
      authorizeRoom({
        user: { sub: "auth0|adjuster", tenant: "org_acme", roles: null },
        room: "org_acme/claims",
      }),
    ).toBe("editor");
    expect(
      authorizeRoom({
        user: { sub: "auth0|auditor", tenant: "org_acme", roles: null },
        room: "org_acme/claims",
      }),
    ).toBe("viewer");
  });

  it("fails closed for missing rooms or subjects when auth is enabled", () => {
    const authorizeRoom = createMembershipRoleResolver({
      memberships: decodeMemberships({ rooms: { claims: { "auth0|u": "owner" } } }),
      defaultRole: null,
    });

    expect(authorizeRoom({ user: { sub: "auth0|other", tenant: null, roles: null }, room: "claims" })).toBeNull();
    expect(authorizeRoom({ user: { sub: "auth0|u", tenant: null, roles: null }, room: "other" })).toBeNull();
  });

  it("keeps tenant isolation even if a cross-tenant membership exists", () => {
    const authorizeRoom = createMembershipRoleResolver({
      memberships: decodeMemberships({
        rooms: {
          "org_evil/claims": { "auth0|adjuster": "owner" },
        },
      }),
    });

    expect(
      authorizeRoom({
        user: { sub: "auth0|adjuster", tenant: "org_acme", roles: null },
        room: "org_evil/claims",
      }),
    ).toBeNull();
  });

  it("can preserve zero-auth dev access through an explicit default role", () => {
    const authorizeRoom = createMembershipRoleResolver({
      memberships: decodeMemberships({ rooms: {} }),
      defaultRole: "editor",
    });

    expect(authorizeRoom({ user: null, room: "playground" })).toBe("editor");
  });

  it("loads memberships from a JSON file", () => {
    const dir = mkdtempSync(join(tmpdir(), "collab-membership-"));
    const file = join(dir, "members.json");
    writeFileSync(file, JSON.stringify({ rooms: { board: { "auth0|owner": "owner" } } }));

    const authorizeRoom = loadMembershipRoleResolver(file);

    expect(authorizeRoom({ user: { sub: "auth0|owner", tenant: null, roles: null }, room: "board" })).toBe(
      "owner",
    );
  });
});

File: modules/relay/relay/membership_test.go
// Ported from modules/collab/test/integration/membership.test.mjs.
package relay

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestDecodeMembershipsStrictRoomSubjectRole(t *testing.T) {
	m, err := DecodeMemberships([]byte(`{"rooms":{"org_acme/claims":{"auth0|adjuster":"editor","auth0|auditor":"viewer"}}}`))
	if err != nil {
		t.Fatalf("decode: %v", err)
	}
	if m["org_acme/claims"]["auth0|adjuster"] != RoleEditor {
		t.Errorf("adjuster role = %q, want editor", m["org_acme/claims"]["auth0|adjuster"])
	}
	if m["org_acme/claims"]["auth0|auditor"] != RoleViewer {
		t.Errorf("auditor role = %q, want viewer", m["org_acme/claims"]["auth0|auditor"])
	}
}

func TestDecodeMembershipsRejectsMalformedData(t *testing.T) {
	cases := []struct {
		name, input, wantErrSubstr string
	}{
		{"invalid role", `{"rooms":{"claims":{"auth0|u":"admin"}}}`, "invalid role"},
		{"members not object", `{"rooms":{"claims":["auth0|u"]}}`, "must be an object"},
		{"no rooms key", `{}`, "rooms object"},
	}
	for _, c := range cases {
		t.Run(c.name, func(t *testing.T) {
			_, err := DecodeMemberships([]byte(c.input))
			if err == nil || !strings.Contains(err.Error(), c.wantErrSubstr) {
				t.Errorf("err = %v, want containing %q", err, c.wantErrSubstr)
			}
		})
	}
}

func TestMembershipGrantsRoleForVerifiedSubject(t *testing.T) {
	m, err := DecodeMemberships([]byte(`{"rooms":{"org_acme/claims":{"auth0|adjuster":"editor","auth0|auditor":"viewer"}}}`))
	if err != nil {
		t.Fatalf("decode: %v", err)
	}
	authorizeRoom := NewMembershipRoleResolver(m, "")
	adjuster := &User{Sub: "auth0|adjuster", Tenant: "org_acme"}
	if got := authorizeRoom(adjuster, "org_acme/claims"); got != RoleEditor {
		t.Errorf("got %q, want editor", got)
	}
	auditor := &User{Sub: "auth0|auditor", Tenant: "org_acme"}
	if got := authorizeRoom(auditor, "org_acme/claims"); got != RoleViewer {
		t.Errorf("got %q, want viewer", got)
	}
}

func TestMembershipFailsClosedForMissingRoomsOrSubjects(t *testing.T) {
	m, err := DecodeMemberships([]byte(`{"rooms":{"claims":{"auth0|u":"owner"}}}`))
	if err != nil {
		t.Fatalf("decode: %v", err)
	}
	authorizeRoom := NewMembershipRoleResolver(m, "")
	if got := authorizeRoom(&User{Sub: "auth0|other"}, "claims"); got != "" {
		t.Errorf("unlisted subject: got %q, want no access", got)
	}
	if got := authorizeRoom(&User{Sub: "auth0|u"}, "other"); got != "" {
		t.Errorf("unlisted room: got %q, want no access", got)
	}
}

func TestMembershipKeepsTenantIsolation(t *testing.T) {
	m, err := DecodeMemberships([]byte(`{"rooms":{"org_evil/claims":{"auth0|adjuster":"owner"}}}`))
	if err != nil {
		t.Fatalf("decode: %v", err)
	}
	authorizeRoom := NewMembershipRoleResolver(m, "")
	user := &User{Sub: "auth0|adjuster", Tenant: "org_acme"}
	if got := authorizeRoom(user, "org_evil/claims"); got != "" {
		t.Errorf("cross-tenant membership: got %q, want no access", got)
	}
}

func TestMembershipPreservesZeroAuthDevAccess(t *testing.T) {
	m, err := DecodeMemberships([]byte(`{"rooms":{}}`))
	if err != nil {
		t.Fatalf("decode: %v", err)
	}
	authorizeRoom := NewMembershipRoleResolver(m, RoleEditor)
	if got := authorizeRoom(nil, "playground"); got != RoleEditor {
		t.Errorf("got %q, want editor", got)
	}
}

func TestLoadMembershipRoleResolverFromFile(t *testing.T) {
	dir := t.TempDir()
	file := filepath.Join(dir, "members.json")
	if err := os.WriteFile(file, []byte(`{"rooms":{"board":{"auth0|owner":"owner"}}}`), 0o644); err != nil {
		t.Fatalf("write file: %v", err)
	}
	authorizeRoom, err := LoadMembershipRoleResolver(file, "")
	if err != nil {
		t.Fatalf("load: %v", err)
	}
	if got := authorizeRoom(&User{Sub: "auth0|owner"}, "board"); got != RoleOwner {
		t.Errorf("got %q, want owner", got)
	}
}
